cephfs/cephfs.go

```diff
@@ -168,3 +263,300 @@ func (mount *MountInfo) IsMounted() bool {
+		log.Errorf("ListDir: Failed to close dir: %s", path)
+		return nil, cephError(ret)
+	}
+
+	return dirs, nil
+}
+
+func (mount *MountInfo) fillCephStat(stx Statx) CephStat {
+	var stat CephStat
+	stat.Mode = uint16(stx.stx.stx_mode)
+	stat.Uid = uint(stx.stx.stx_uid)
+	stat.Gid = uint(stx.stx.stx_gid)
+	stat.Size = uint64(stx.stx.stx_size)
+	stat.BlkSize = uint(stx.stx.stx_blksize)
+	stat.Blocks = uint64(stx.stx.stx_blocks)
+
+	time := stx.stx.stx_atime.tv_sec
+	time *= 1000
+	time += stx.stx.stx_atime.tv_nsec / 1000000
+	stat.AppendTime = int64(time)
+
+	time = stx.stx.stx_mtime.tv_sec
+	time *= 1000
+	time += stx.stx.stx_mtime.tv_nsec / 1000000
+	stat.ModifyTime = int64(time)
+
+	if (stx.stx.stx_mode & C.S_IFMT) == C.S_IFREG {
+		stat.IsFile = true
+	} else {
+		stat.IsFile = false
+	}
+
+	if (stx.stx.stx_mode & C.S_IFMT) == C.S_IFDIR {
+		stat.IsDir = true
+	} else {
+		stat.IsDir = false
+	}
+
+	if (stx.stx.stx_mode & C.S_IFMT) == C.S_IFLNK {
+		stat.IsSymlink = true
+	} else {
+		stat.IsSymlink = false
+	}
+	return stat
+}
+
+// Stat get file status from a given path
+func (mount *MountInfo) Stat(path string) (CephStat, error) {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	stx := Statx{}
+	ret := C.ceph_statx(mount.mount, cPath, &stx.stx, CephStatMask, 0)
+	// ret := C.ceph_statx(mount.mount, cPath, &stx.stx, C.CEPH_STATX_UID, 0)
+	if ret != 0 {
+		log.Errorf("Stat: Failed to get file status: %s", path)
+		return CephStat{}, cephError(ret)
+	}
+
+	stat := mount.fillCephStat(stx)
+	return stat, nil
+}
+
+// FStat get file status from a file descriptor
+func (mount *MountInfo) FStat(fd int) (CephStat, error) {
+	stx := Statx{}
+	ret := C.ceph_fstatx(mount.mount, C.int(fd), &stx.stx, CephStatMask, 0)
+	if ret != 0 {
+		log.Errorf("Stat: Failed to get file status")
+		return CephStat{}, cephError(ret)
+	}
+
+	stat := mount.fillCephStat(stx)
+	return stat, nil
+}
+
+// LStat get file status, without following symlinks
+func (mount *MountInfo) LStat(path string) (CephStat, error) {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	stx := Statx{}
+	ret := C.ceph_statx(mount.mount, cPath, &stx.stx, CephStatMask, C.AT_SYMLINK_NOFOLLOW)
+	if ret != 0 {
+		log.Errorf("Stat: Failed to get file status: %s", path)
+		return CephStat{}, cephError(ret)
+	}
+
+	stat := mount.fillCephStat(stx)
+	return stat, nil
+}
+
+/*
+// SetAttr set file attributes
+func (mount *MountInfo) SetAttr(path string, stat CephStat, mask int) {
+
+}
+*/
+// Truncate truncate a file to a specified length
+func (mount *MountInfo) Truncate(path string, size int64) error {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	ret := C.ceph_truncate(mount.mount, cPath, C.int64_t(size))
+	if ret != 0 {
+		log.Errorf("Truncate: Failed to truncate file: %s", path)
+		return cephError(ret)
+	}
+	return nil
+}
+
+// FTruncate truncate a file to a specified length
+func (mount *MountInfo) FTruncate(fd int, size int64) error {
```

Comment:
Tests for truncate/ftruncate too

Reply:
Hi noah, I have added the test for truncate/ftruncate/lseek/fchmod, it should work now